ci/src/main/java/ci/ContinuousIntegrationServer.java

```diff
@@ -25,81 +25,138 @@
-                break;
+        String url = request.getRequestURL().toString();
+
+        String subDirectory = "";
+        boolean singleBuild = false;
+        int ifSubDirectoryURL = url.indexOf('/', 8);
+        // true if URL matches http(s)://<host>/*, false if matches http(s)://<host>
+        if (ifSubDirectoryURL != -1) {
+            subDirectory = url.substring(ifSubDirectoryURL, url.length());
+            // use regex to see if any specific build is accessed
+            Pattern singleBuildPattern = Pattern.compile("/builds/.");
+            Matcher matcher = singleBuildPattern.matcher(subDirectory);
+            // true if URL matches regex of singleBuildPattern
+            singleBuild = matcher.find();
         }
+        
+        // show list of all stored builds if URL is http(s)://<host>/builds
+        if (subDirectory.equals("/builds")) {
+            response.setContentType("text/html");
+
+            // get all build filenames in buildHistory directory
+            File dir = new File("../buildHistory");
+            String[] buildFileNames = dir.list();
+            // start of simple html file
+            String output = "<!DOCTYPE html>\n<html>\n<body>\n<h1>Build list</h1>";
+            
+            // iterate through each build
+            for (String build : buildFileNames) {
+                String buildURL = url + "/" + build;
+                // add link to the build in the html
+                output += "<a href=\"" + buildURL + "\">" + build + "</a> <br>";
+            }
 
+            output += "</body>\n</html>"; // end of simple html file
+            response.getWriter().print(output);
+            response.setStatus(HttpServletResponse.SC_OK);
+            baseRequest.setHandled(true);
+        }
+        // show single build information if URL is http(s)://<host>/builds/<singleBuildFileName>
+        else if (singleBuild) {
+            String output = "";
 
+            // get the name of the build from the URL
+            int indexOfBuilds = url.indexOf("builds/");
+            String buildName = url.substring(indexOfBuilds + 7, url.length());
 
+            try {
+                File build = new File("../buildHistory/" + buildName);
+                BufferedReader br = new BufferedReader(new FileReader(build));
+                String line;
+                while ((line = br.readLine()) != null) {
+                    output += line + "\n";
+                }
+                br.close();
+            }
+            catch (Exception e) {
+                output += "error: no such build exists";
+            }
 
-        response.setContentType("text/html;charset=utf-8");
-        response.setStatus(HttpServletResponse.SC_OK);
-        baseRequest.setHandled(true);
-
-
-
-
-
+            response.getWriter().print(output);
+            response.setStatus(HttpServletResponse.SC_OK);
+            baseRequest.setHandled(true);
+        }
+        else {
```

Comment:
will we not miss any push if it's in the if-else statement or are the build information separated from the push events?

Reply:
retrieving the build information has nothing to do with the github webhook. it is a separate request to the CI server.